gemrb/core/Scriptable/Actor.cpp

```diff
@@ -5439,15 +5435,42 @@ int Actor::GetEncumbranceFactor(bool feedback) const
 // NOTE: for ini-based speed users this will only update their encumbrance, speed will be 0
 int Actor::CalculateSpeed(bool feedback)
 {
-	int speed = GetStat(IE_MOVEMENTRATE);
+	int movementRate = GetStat(IE_MOVEMENTRATE);
+	int encumbranceFactor = GetEncumbranceFactor(feedback);
 
 	if (BaseStats[IE_EA] > EA_GOODCUTOFF && !third) {
 		// cheating bastards (drow in ar2401 for example)
-		return speed;
+	} else {
+		movementRate /= encumbranceFactor;
 	}
-
-	int encumbranceFactor = GetEncumbranceFactor(feedback);
-	return speed / encumbranceFactor;
+	if (movementRate) {
+		speed = 1500 / movementRate;
+	} else {
+		speed = 0;
+	}
+	if (core->HasFeature(GF_RESDATA_INI)) {
```

Comment:
This block is completely separate even in Map. Since the function does two unrelated things at once, the proper thing is two add two more private functions and have CalculateSpeed basically be:
```
if (core->HasFeature(GF_RESDATA_INI)) {
  CalculateSpeedFromIni(feedback);
} else {
  CalculateSpeedFromRate(feedback);
}
```